modules/Bio/EnsEMBL/Variation/Pipeline/AncestralAlleles/Assign.pm

```diff
@@ -0,0 +1,168 @@
+  my $batch = $self->param('batch');
+  my $id = $batch->{id};
+  my $start = $batch->{start};
+  my $end = $batch->{end};
+  my $ancestral_file = $self->param('ancestral_file');
+
+  $self->setup();
+  $self->dump_variation_features();
+  $self->assign_ancestral_alleles();
+  $self->cleanup();
+}
+
+sub setup {
+  my $self = shift;
+  my $batch = $self->param('batch');
+  my $id = $batch->{id};
+  my $compara_dir = $self->param('compara_dir'); 
+  my $species_dir = $self->param('species_dir');
+  my $ancestral_file_targz = $self->param('ancestral_file');
+  my $ancestral_file_notargz = $ancestral_file_targz;
+  $ancestral_file_notargz =~ s/\.tar\.gz//;
+  # create temp directory for storing and uncompressing ancestral fasta file 
+  unless (-d "$species_dir/fasta_$id") {
+    my $err;
+    make_path("$species_dir/fasta_$id", {error => \$err});
+    die "make_path failed: ".Dumper($err) if $err && @$err;
+  }
+  # cp  ancestral fasta file from compara dir to temp directory
+  $self->run_cmd("cp $compara_dir/$ancestral_file_targz $species_dir/fasta_$id");
+  # uncompress ancestral fasta file
+  $self->run_cmd("tar xzf $species_dir/fasta_$id/$ancestral_file_targz -C $species_dir/fasta_$id/");
+  # rm bed files
+  $self->run_cmd("rm $species_dir/fasta_$id/$ancestral_file_notargz/*.bed");
+  my $fasta_files_dir = "$species_dir/fasta_$id/$ancestral_file_notargz"; 
+  $self->param('fasta_files_dir', $fasta_files_dir);
+  my $fasta_files_dir_root = "$species_dir/fasta_$id/"; 
+  $self->param('fasta_files_dir_root', $fasta_files_dir_root);
+}
+
+sub dump_variation_features {
+  my $self = shift;
+  my $species_name = $self->param('species_name');
+  my $species_dir = $self->param('species_dir');
+  my $batch = $self->param('batch'); 
+  my ($start, $end) = ($batch->{start}, $batch->{end});
+  my $vf_file = "$species_dir/vf_$start\_$end";
+  my $aa_file = "$species_dir/aa_$start\_$end";
+  $self->param('vf_file', $vf_file);
+  $self->param('aa_file', $aa_file);
+  if (! -e "$vf_file") {
+    my $fh = FileHandle->new($vf_file, 'w');
+    $self->param('species', $species_name);
+    my $vdba = $self->get_species_adaptor('variation');
+    my $dbc = $vdba->dbc;
+    my $sth = $dbc->prepare(qq{
+      SELECT vf.variation_feature_id, sr.name, vf.seq_region_start, vf.seq_region_end, vf.seq_region_strand, vf.ancestral_allele
+      FROM variation_feature vf, seq_region sr
+      WHERE vf.seq_region_id = sr.seq_region_id
+      AND vf.variation_feature_id >= ?
+      AND vf.variation_feature_id <= ?
+      ORDER BY vf.variation_feature_id;}, {mysql_use_result => 1}
+    ) or die $dbc->errstr;
+    $sth->execute($start, $end) or die $dbc->errstr;
+    my ($vf_id, $seq_region, $start, $end, $strand, $aa);
+    $sth->bind_columns(\($vf_id, $seq_region, $start, $end, $strand, $aa));
+    while ($sth->fetch) {
+      $aa ||= 'NULL';
+      print $fh join("\t", ($vf_id, $seq_region, $start, $end, $strand, $aa)), "\n";
```

Comment:
It should be quicker to calculate the AA without writing the input to disk and reading it again.

Reply:
This is a precautionary measure in case we have to deal with any failures and rerun the job. Then we don't need to connect to the database again but can annotate from file. But I agree that it is a manageable size to annotate on the fly.